### PR TITLE
[Feature] API Show New User Profile Fields

### PR DESCRIPTION
**Pull request by billyhunt**
Fixes bug #4203

Purpose
-----------
In the API, users that have not filled out the any social fields in their profile will not show social media fieldnames. There is just a blank dictionary labeled social_accounts.

screen shot 2015-08-28 at 9 34 28 am

To provide consistent user experience and to encourage users to fill out their profiles, we will now show all social fields even if they do not exist on the database.  

Changes
------------
User fields are shown on new users and ones without any social media information filled out.

Side Effects
---------------
All User social fields are stored at the top level of the JSON object, not one level down in the social_accounts dictionary.

#### Before:  

screen shot 2015-08-28 at 9 52 14 am

#### Now:

screen shot 2015-08-28 at 9 37 42 am

### CODE REVIEW

File: api/users/serializers.py [hunk 1]
```diff
@@ -1,6 +1,87 @@
 from rest_framework import serializers as ser
 from api.base.serializers import JSONAPISerializer, LinksField, Link
 from website.models import User
+from rest_framework.fields import empty, SkipField, get_attribute, CharField, MaxLengthValidator, MinLengthValidator, URLValidator, URLField
+
+
+class SocialCharField(ser.CharField):
+
+    def __init__(self, social_default, **kwargs):
+        self.allow_blank = kwargs.pop('allow_blank', False)
```
Comment (brianjgeiger) [inline api/users/serializers.py]: The last line of this you can keep, but you should be able to super(SocialCharField, self).__init__(**kwargs) after it and not have to use the rest of the init method, since all of that is covered by the superclass.

File: api/users/serializers.py [hunk 2]
```diff
@@ -1,6 +1,87 @@
 from rest_framework import serializers as ser
 from api.base.serializers import JSONAPISerializer, LinksField, Link
 from website.models import User
+from rest_framework.fields import empty, SkipField, get_attribute, CharField, MaxLengthValidator, MinLengthValidator, URLValidator, URLField
+
+
+class SocialCharField(ser.CharField):
```
Comment (brianjgeiger) [inline api/users/serializers.py]: Let's name this a bit more generically. Something like `CharFieldWithReadDefault` and change the variable to match. We don't need to go too crazy with the descriptive name, but put in the docstring what the purpose of this field is (i.e. handle if a dictionary is being accessed for the field but the field needs to return a default value).

File: api/users/serializers.py [hunk 3]
```diff
@@ -1,6 +1,87 @@
...
+        self.max_length = kwargs.pop('max_length', None)
+        self.min_length = kwargs.pop('min_length', None)
+        super(CharField, self).__init__(**kwargs)
+        if self.max_length is not None:
+            message = self.error_messages['max_length'].format(max_length=self.max_length)
+            self.validators.append(MaxLengthValidator(self.max_length, message=message))
+        if self.min_length is not None:
+            message = self.error_messages['min_length'].format(min_length=self.min_length)
+            self.validators.append(MinLengthValidator(self.min_length, message=message))
+        self.social_default = social_default
```
Comment (brianjgeiger) [inline api/users/serializers.py]: Hmmm. It's not that we want a default value, really. What we want is to make sure that the field always shows even if it's empty. What about making this `self.show_field_when_empty` and make it a boolean.

File: api/users/serializers.py [hunk 4]
```diff
@@ -1,6 +1,87 @@
...
+        """
+        Given the *outgoing* object instance, return the primitive value
+        that should be used for this field.
+        """
+        try:
+            return get_attribute(instance, self.source_attrs)
+        except (KeyError, AttributeError) as exc:
+            if not self.required and self.default and self.social_default is empty:
+                raise SkipField()
+            if self.social_default == '':
```
Comment (brianjgeiger) [inline api/users/serializers.py]: With the `self.always_show` boolean just check to make sure it's true.

File: api/users/serializers.py [hunk 5]
```diff
@@ -1,6 +1,87 @@
...
+            message = self.error_messages['min_length'].format(min_length=self.min_length)
+            self.validators.append(MinLengthValidator(self.min_length, message=message))
+        self.social_default = social_default
+
+    def get_attribute(self, instance):
+        """
+        Given the *outgoing* object instance, return the primitive value
+        that should be used for this field.
+        """
+        try:
```
Comment (brianjgeiger) [inline api/users/serializers.py]: It might be good to check if self.always_show is true here, and if not, call the superclass to process this function. That will ensure that, if things change on the CharField, you're not duplicating that unnecessarily. A minor point and could be argued, because I don't know if you're going to get away from being able to duplicate most of the code anyways.

File: api/users/serializers.py [hunk 6]
```diff
@@ -1,6 +1,87 @@
...
+            raise type(exc)(msg)
+
+
+class SocialUrlField(ser.URLField):
+
+    def __init__(self, social_default, **kwargs):
+        super(URLField, self).__init__(**kwargs)
+        validator = URLValidator(message=self.error_messages['invalid'])
+        self.validators.append(validator)
+        self.social_default = social_default
```
Comment (brianjgeiger) [inline api/users/serializers.py]: Same comments as above. DRY, rename field, etc.

**Comment by brianjgeiger** [comment]
First pass done. :penguin: Looks good, but make it a bit more generic and a little more DRY, but the general approach is great and should do just what we want.

**Comment by billyhunt** [comment]
In the current API implementation, when updating a field in the social dictionary, it is the only field written.  So, for example, when you patch a new user's github field, it will only add 'github':'newGitHub" to the social dictionary, but not add 'personal': '' or any other social key.  This is different then the OSF, which writes all of the keys to the social dictionary when any field is updated.

The blank fields are created by the serializer in the API (the reason for this PR), but only the modified keys and values are written to the DB.

So, a PATCH or PUT has different behavior then currently exists on the osf.  No tests fail, all fields write correctly, and my functional testing has turned up no errors on the osf or the api.  We could add the missing blank fields when we do any write (like the OSF), but that seems not 100% inline with the DRF specifications.  @sloria, @brianjgeiger, thoughts?

Make sense?

**Comment by brianjgeiger** [comment]
I think keep it as you have it. I don't see the need to replicate OSF functionality in the API.

**Comment by brianjgeiger** [comment]
Looks good to me. :tada: :penguin: :tada: Looks ready for @sloria.

**Comment by HarryRybacki** [comment]
[#OSF-4257](https://openscience.atlassian.net/browse/OSF-4257)

File: api/users/serializers.py [hunk 7]
```diff
@@ -1,6 +1,37 @@
...
+
+
+class CharFieldWithReadDefault(ser.CharField):
+
+    def get_attribute(self, instance):
+        """
+        Overwrite the error message to return a blank value is if there is no existing value.
+        This allows the display of keys that do not exist in the DB (gitHub on a new OSF account for example.)
+        """
+        try:
```
Comment (sloria) [inline api/users/serializers.py]: This is almost identical to the parent class's implementation. Instead of copying the code, you could call the parent class's `get_attribute` and handle the `KeyError` and `AttributeError`.

File: api/users/serializers.py [hunk 8]
```diff
@@ -1,6 +1,37 @@
...
+
+
+class UrlFieldWithReadDefault(ser.URLField):
+
+    def get_attribute(self, instance):
+        """
+        Overwrite the error message to return a blank value is if there is no existing value.
+        This allows the display of keys that do not exist in the DB (profile_website on a new OSF account for example.)
+        """
+        try:
```
Comment (sloria) [inline api/users/serializers.py]: See previous comment.

File: api/users/serializers.py [hunk 9]
```diff
@@ -1,6 +1,37 @@
 from rest_framework import serializers as ser
 from api.base.serializers import JSONAPISerializer, LinksField, Link
 from website.models import User
+from rest_framework.fields import SkipField, get_attribute
+
+
+class CharFieldWithReadDefault(ser.CharField):
```
Comment (sloria) [inline api/users/serializers.py]: These new fields are specific to users; therefore, they belong in `api.base.serializers`.

**Comment by sloria** [comment]
WIth the current approach, you would have to make a subclass for every field that you want to allow a missing attribute. This is not ideal.

How about using composition instead?

```python
twitter = AllowMissing(CharField(source='social.twitter', ...))
```

**Comment by sloria** [comment]
Pass finished.

**Comment by billyhunt** [comment]
There seems to be some django rest inheritance issues with using a composition.  Since this is a specific use case, I think overriding the two DRF classes to test for blanks is the best solution.

**Comment by sloria** [comment]
What, specifically, are the problems?

**Comment by billyhunt** [comment]
The error I am getting is 

My composed class is running the Field class' to_representation function, when it should be running the specific field's (CharField) to_representation.  

Here is the specific error message.

```
AllowMissing.to_representation() must be implemented.
If you are upgrading from REST framework version 2 you might want `ReadOnlyField`.
```

and my implementation

```
class AllowMissing(ser.Field):

    def __init__(self, *args, **kwargs):
        super(AllowMissing, self).__init__(*args, **kwargs)

    def get_attribute(self, instance):
        """
        Overwrite the error message to return a blank value is if there is no existing value.
        This allows the display of keys that do not exist in the DB (gitHub on a new OSF account for example.)
        """
        try:
            Field.get_attribute(self, instance)
        except SkipField:
            return ''
```

**Comment by sloria** [comment]
You should probably follow the advice in that error message. =) It's telling you that fields subclasses must implement `to_representation`.

File: scripts/migration/migrate_personal_to_profile_websites.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+from tests.factories import UserFactory
+
+from website import models
+from website.app import init_app
+from scripts import utils as scripts_utils
+
+logger = logging.getLogger(__name__)
+
+
+def main():
```
Comment (sloria) [inline scripts/migration/migrate_personal_to_profile_websites.py]: Is this script related to this PR? Seems to it might be for a different feature.

File: api/base/serializers.py [hunk 11]
```diff
@@ -12,6 +12,33 @@
 from website.util import waterbutler_api_url_for
 
 
+class AllowMissing(ser.Field):
```
Comment (sloria) [inline api/base/serializers.py]: Please add a docstring explaining the purpose of this field.

File: api/base/serializers.py [hunk 12]
```diff
@@ -273,9 +300,10 @@ def to_representation(self, obj, envelope='data'):
                 attribute = field.get_attribute(obj)
             except SkipField:
                 continue
-
             if isinstance(field, JSONAPIHyperlinkedIdentityField):
                 data['relationships'][field.field_name] = field.to_representation(attribute)
+            # if isinstance(field, AllowMissing):
```
Comment (sloria) [inline api/base/serializers.py]: Remove these lines.

**Comment by sloria** [comment]
Just a few last minor changes.

**Comment by billyhunt** [comment]
Replaced by #4357
